tracer/src/Datadog.Trace/AppSec/WafEncoding/Encoder.cs

```diff
@@ -63,442 +63,459 @@ public static string FormatArgs(object o)
                     break;
-                }
-
                 case decimal d:
-                {
                     ddwafObjectStruct = new DdwafObjectStruct { Type = DDWAF_OBJ_TYPE.DDWAF_OBJ_DOUBLE, DoubleValue = (double)d };
                     break;
-                }
-
                 case double d:
-                {
                     ddwafObjectStruct = new DdwafObjectStruct { Type = DDWAF_OBJ_TYPE.DDWAF_OBJ_DOUBLE, DoubleValue = d };
                     break;
-                }
-
                 case float d:
-                {
                     ddwafObjectStruct = new DdwafObjectStruct { Type = DDWAF_OBJ_TYPE.DDWAF_OBJ_DOUBLE, DoubleValue = d };
                     break;
-                }
-
                 case bool b:
                     ddwafObjectStruct = new DdwafObjectStruct { Type = DDWAF_OBJ_TYPE.DDWAF_OBJ_BOOL, ByteValue = b ? (byte)1 : (byte)0 };
                     break;
-
                 case IEnumerable<KeyValuePair<string, object>> objDict:
-                {
-                    var collectionDict = objDict as ICollection<KeyValuePair<string, object>> ?? objDict.ToList();
-                    var count = collectionDict.Count;
-                    ddwafObjectStruct = ProcessKeyValuePairs(collectionDict, count, &GetKey1, &GetValue1);
-                    static string GetKey1(KeyValuePair<string, object> item) => item.Key;
-                    static object GetValue1(KeyValuePair<string, object> item) => item.Value;
-                    break;
-                }
+                    {
+                        var collectionDict = objDict as ICollection<KeyValuePair<string, object>> ?? objDict.ToList();
+                        ddwafObjectStruct = ProcessKeyValuePairs(ref context, remainingDepth, key, collectionDict, collectionDict.Count, &GetKey1, &GetValue1);
+                        static string GetKey1(KeyValuePair<string, object> item) => item.Key;
+                        static object GetValue1(KeyValuePair<string, object> item) => item.Value;
+                        break;
+                    }
 
                 case IEnumerable<KeyValuePair<string, bool>> objDict:
-                {
-                    var collectionDict = objDict as ICollection<KeyValuePair<string, bool>> ?? objDict.ToList();
-                    var count = collectionDict.Count;
-                    ddwafObjectStruct = ProcessKeyValuePairs(collectionDict, count, &GetKey1, &GetValue1);
-                    static string GetKey1(KeyValuePair<string, bool> item) => item.Key;
-                    static object GetValue1(KeyValuePair<string, bool> item) => item.Value;
-                    break;
-                }
+                    {
+                        var collectionDict = objDict as ICollection<KeyValuePair<string, bool>> ?? objDict.ToList();
+                        ddwafObjectStruct = ProcessKeyValuePairs(ref context, remainingDepth, key, collectionDict, collectionDict.Count, &GetKey1, &GetValue1);
```

Comment:
NIT: wondering if `remainingDepth` could be part of a mutable field in the `EncoderContext` 🤔